File: syncloud/src/main/java/org/syncloud/android/ui/DevicesDiscoveryActivity.java
package org.syncloud.android.ui;

import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.net.nsd.NsdManager;
import android.net.wifi.WifiManager;
import android.os.AsyncTask;
import android.os.Bundle;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.appcompat.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import com.google.common.base.Optional;

import org.apache.log4j.Logger;
import org.syncloud.android.Preferences;
import org.syncloud.android.R;
import org.syncloud.android.SyncloudApplication;
import org.syncloud.android.core.platform.model.Identification;
import org.syncloud.android.discovery.DeviceEndpointListener;
import org.syncloud.android.discovery.DiscoveryManager;
import org.syncloud.android.ui.adapters.DevicesDiscoveredAdapter;
import org.syncloud.android.ui.dialog.WifiDialog;
import org.syncloud.android.core.platform.Internal;
import org.syncloud.android.core.platform.model.Endpoint;
import org.syncloud.android.core.platform.model.IdentifiedEndpoint;

import java.util.Map;

import static com.google.common.collect.Maps.newHashMap;

public class DevicesDiscoveryActivity extends AppCompatActivity {

    private static Logger logger = Logger.getLogger(DevicesDiscoveryActivity.class.getName());

    private Preferences preferences;

    private DiscoveryManager discoveryManager;
    private FloatingActionButton refreshBtn;
    private DevicesDiscoveredAdapter listAdapter;
    private SwipeRefreshLayout swipeRefreshLayout;
    private View emptyView;

    private ListView resultsList;

    private Map<Endpoint, IdentifiedEndpoint> map;
    private Internal internal;
    private SyncloudApplication application;

    private static int REQUEST_SETTINGS = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        setContentView(R.layout.activity_devices_discovery);

        application = (SyncloudApplication) getApplication();
        preferences = application.getPreferences();
        internal = new Internal();

        swipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.swipe_refresh_layout);
        swipeRefreshLayout.setColorSchemeResources(R.color.logo_blue, R.color.logo_green);
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                checkWiFiAndDiscover();
            }
        });

        emptyView = findViewById(android.R.id.empty);

        resultsList = (ListView) findViewById(R.id.devices_discovered);

        refreshBtn = (FloatingActionButton) findViewById(R.id.discovery_refresh_btn);
        listAdapter = new DevicesDiscoveredAdapter(this);
        resultsList.setAdapter(listAdapter);

        resultsList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int position, long l) {
                Object obj = resultsList.getItemAtPosition(position);
                IdentifiedEndpoint ie = (IdentifiedEndpoint) obj;
                open(ie);
            }
        });

        map = newHashMap();

        discoveryManager = new DiscoveryManager(
                (WifiManager) getSystemService(Context.WIFI_SERVICE),
                (NsdManager) getSystemService(Context.NSD_SERVICE));

        swipeRefreshLayout.post(new Runnable() {
            @Override
            public void run() {
                checkWiFiAndDiscover();
            }
        });
    }

    private void checkWiFiAndDiscover() {
        listAdapter.clear();
        if (application.isWifiConnected()) {
            new DiscoveryTask().execute();
        } else {
            WifiDialog dialog = new WifiDialog();
            dialog.setMessage("Discovery is possible only in the same Wi-Fi network where you have Syncloud device connected.");
            dialog.show(getSupportFragmentManager(), "discovery_wifi_dialog");
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == WifiDialog.WIFI_SETTINGS) {
            checkWiFiAndDiscover();
        } else {
            finish();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            Intent intentSettings = new Intent(this, SettingsActivity.class);
            startActivityForResult(intentSettings, REQUEST_SETTINGS);
        }
        return super.onOptionsItemSelected(item);
    }

    private void open(final IdentifiedEndpoint endpoint) {
        if (!endpoint.id().isPresent()) {
            new AlertDialog.Builder(this)
                    .setTitle("Can't identify device")
                    .setMessage("Sorry, there's no identification information for this device. Most probably it is running old release of Syncloud. Please upgrade it to latest release and try to activate again.")
                    .setPositiveButton("OK", null)
                    .show();
        } else {

            Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(endpoint.endpoint().activationUrl()));
            startActivity(browserIntent);

        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        logger.info("leaving the screen");
        discoveryManager.cancel();
    }

    public void refresh(View view) {
        checkWiFiAndDiscover();
    }

    public class DiscoveryTask extends AsyncTask<Void, Progress, Void> {

        private final DeviceEndpointListener deviceEndpointListener;

        public DiscoveryTask() {
            deviceEndpointListener = new DeviceEndpointListener() {
                @Override
                public void added(final Endpoint endpoint) {
                    Optional<Identification> id = internal.getId(endpoint.host());
                    if (id.isPresent()) {
                        IdentifiedEndpoint ie = new IdentifiedEndpoint(endpoint, id);
                        publishProgress(new Progress(true, endpoint, ie));
                    }
                }

                @Override
                public void removed(final Endpoint endpoint) {
                    IdentifiedEndpoint ie = map.remove(endpoint);
                    publishProgress(new Progress(false, endpoint, ie));
                }
            };
        }

        @Override
        protected void onPreExecute() {
            refreshBtn.setVisibility(View.GONE);
            swipeRefreshLayout.setRefreshing(true);
            emptyView.setVisibility(View.GONE);
            resultsList.setEmptyView(null);
            listAdapter.clear();

            //use for testing without wi-fi
            //listAdapter.add(new DirectEndpoint("localhost", 22, "vsapronov", "somepassword", null));
        }

        @Override
        protected Void doInBackground(Void... params) {
            discoveryManager.run(20, deviceEndpointListener);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            emptyView.setVisibility(View.VISIBLE);
            resultsList.setEmptyView(emptyView);
            swipeRefreshLayout.setRefreshing(false);
            refreshBtn.setVisibility(View.VISIBLE);
        }

        @Override
        protected void onProgressUpdate(Progress... progresses) {
            Progress progress = progresses[0];
            IdentifiedEndpoint ie = progress.identifiedEndpoint;
            if (progress.isAdded) {
                map.put(progress.endpoint, ie);
                listAdapter.add(ie);
            } else {
                listAdapter.remove(ie);
            }
        }
    }

    public class Progress {
        boolean isAdded = true;
        Endpoint endpoint;
        IdentifiedEndpoint identifiedEndpoint;

        public Progress(boolean isAdded, Endpoint endpoint, IdentifiedEndpoint identifiedEndpoint) {
            this.isAdded = isAdded;
            this.endpoint = endpoint;
            this.identifiedEndpoint = identifiedEndpoint;
        }
    }
}

File: syncloud/src/main/java/org/syncloud/android/ui/DevicesDiscoveryActivity.kt
package org.syncloud.android.ui

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.net.nsd.NsdManager
import android.net.wifi.WifiManager
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.ListItem
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.syncloud.android.R
import org.syncloud.android.SyncloudApplication
import org.syncloud.android.core.common.WebService
import org.syncloud.android.core.common.http.HttpClient
import org.syncloud.android.core.platform.Internal
import org.syncloud.android.core.platform.model.IdentifiedEndpoint
import org.syncloud.android.discovery.DiscoveryManager
import org.syncloud.android.ui.theme.SyncloudTheme

const val DISCOVERY_TIMEOUT_SECONDS = 20

class DevicesDiscoveryActivity : ComponentActivity() {
    private lateinit var discoveryManager: DiscoveryManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        val application = application as SyncloudApplication
        discoveryManager = DiscoveryManager(
            applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager,
            applicationContext.getSystemService(Context.NSD_SERVICE) as NsdManager
        )
        setContent {
            SyncloudTheme {
                DevicesDiscoveryScreen(
                    discoveryManager = discoveryManager,
                    platform = Internal(WebService(HttpClient()))
                )
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        discoveryManager.cancel()
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun DevicesDiscoveryScreen(
    discoveryManager: DiscoveryManager,
    platform: Internal
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val endpoints = remember { mutableStateListOf<IdentifiedEndpoint>() }
    var searching by remember { mutableStateOf(false) }
    var finished by remember { mutableStateOf(false) }
    suspend fun discover() {
        endpoints.clear()
        finished = false
        searching = true
        withContext(Dispatchers.IO) {
            discoveryManager.run(DISCOVERY_TIMEOUT_SECONDS) { device ->
                val id = platform.getId(device)
                if (id != null) {
                    withContext(Dispatchers.Main) {
                        if (endpoints.none { it.device == device }) {
                            endpoints.add(IdentifiedEndpoint(device, id))
                        }
                    }
                }
            }
        }
        searching = false
        finished = true
    }

    LaunchedEffect(Unit) { discover() }

    Scaffold(
        topBar = { TopAppBar(title = { Text(stringResource(R.string.title_activity_discovery)) }) },
        floatingActionButton = {
            FloatingActionButton(
                onClick = { scope.launch { discover() } },
                modifier = Modifier.testTag("discovery_refresh_button")
            ) {
                Icon(
                    painter = painterResource(R.drawable.ic_refresh_white_24dp),
                    contentDescription = stringResource(R.string.refresh_button)
                )
            }
        }
    ) { contentPadding ->
        PullToRefreshBox(
            isRefreshing = searching,
            onRefresh = { scope.launch { discover() } },
            modifier = Modifier
                .fillMaxSize()
                .padding(contentPadding)
                .testTag("devices_discovery_screen")
        ) {
            if (endpoints.isEmpty() && finished) {
                Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                    Text(
                        text = stringResource(R.string.no_devices_found),
                        textAlign = TextAlign.Center,
                        modifier = Modifier
                            .padding(32.dp)
                            .testTag("devices_discovery_empty")
                    )
                }
            }
            LazyColumn(Modifier.fillMaxSize()) {
                items(endpoints) { endpoint ->
                    ListItem(
                        headlineContent = { Text(endpoint.id.title ?: endpoint.device) },
                        supportingContent = { Text(endpoint.device) },
                        modifier = Modifier
                            .clickable {
                                context.startActivity(
                                    Intent(Intent.ACTION_VIEW, Uri.parse("https://${endpoint.device}"))
                                )
                            }
                            .testTag("discovered_item")
                    )
                    HorizontalDivider()
                }
            }
        }
    }
}
